### PR TITLE
tidy up optimization transform testing

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
While working on #4831, we grew suspicious that similar bugs might exist in other optimization transforms. This PR is the result of that audit.

**Description of the Change:**
- make sure optimization-transform tests actually test QNode transforms
  - this involved changing some observables to get actual non-zero results
- simplify the `remove_barrier` code
- add a test for `remove_barrier` (in test_compile.py because the qml.compile default pipeline is the only place where it's used) because there are no tests for it!

**Benefits:**
More comfort with the state of this module.

**Possible Drawbacks:**
N/A

[sc-50172]

### CODE REVIEW

File: pennylane/transforms/optimization/pattern_matching.py [hunk 1]
```diff
@@ -277,8 +277,7 @@ def circuit():
                         qml.apply(inst)
 
                 qscript = QuantumScript.from_queue(q_inside)
-                tapes, fn = qml.map_wires(input=qscript, wire_map=inverse_wires_map)
-                tape = fn(tapes)
```
Comment (timmysilv) [inline pennylane/transforms/optimization/pattern_matching.py:281]: this technically worked but it felt wrong - so I changed it

File: pennylane/transforms/optimization/remove_barrier.py [hunk 2]
```diff
@@ -78,19 +78,7 @@ def qfunc(x, y):
...
-        current_gate = list_copy[0]
-
-        # Remove Barrier gate
-        if current_gate.name != "Barrier":
-            operations.append(current_gate)
-
-        list_copy.pop(0)
-        continue
-
+    operations = filter(lambda op: op.name != "Barrier", tape.operations)
```
Comment (albi3ro) [inline pennylane/transforms/optimization/remove_barrier.py:81]: 🎉

**Review by albi3ro — APPROVED** [review]
Thanks for looking into this.

**Review by lillian542 — APPROVED** [review]
Looks good 🥳

**Pull request merged**